### PR TITLE
fix: lets detect git/credentials provided by tekton

### PR DESCRIPTION
**Pull request by jstrachan**
fixes https://github.com/jenkins-x/jx/issues/6031

### CODE REVIEW

File: pkg/auth/gitcreds/gitcredentials_auth.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+	"github.com/jenkins-x/jx/pkg/auth"
+	"github.com/jenkins-x/jx/pkg/log"
+	"github.com/jenkins-x/jx/pkg/util"
+	"github.com/pkg/errors"
+)
+
+// LoadGitCredentialsAuth loads the git credentials from the `git/credentials` file
+// in `$XDG_CONFIG_HOME/git/credentials` or in the `~/git/credentials` directory
+func LoadGitCredentialsAuth() (*auth.AuthConfig, error) {
```
Comment (ccojocar) [inline pkg/auth/gitcreds/gitcredentials_auth.go]: @jstrachan Have you seen https://github.com/jenkins-x/jx/pull/6034/files? We should abstract this behind a `ConfigHandler` interface which is used by the `ConfigService` to load the config.

**Comment by ccojocar** [comment]
@jstrachan I am wondering what's the advantage to use this approach since we configured the git credentials in boot https://github.com/jenkins-x/jenkins-x-boot-config/blob/7d088ae938eae89fb976d763e1fe47d65038a8f6/env/tekton/values.tmpl.yaml#L14 and are the same with pipeline user/token from vault.

**Comment by ccojocar** [comment]
Also we are working to an overall improvement for sects management. See https://github.com/cloudbees/jenkins-x-security/issues/12. 

The list of issues assigned to this topic https://codetree.com/projects/StX1/issues/10427820?type=epic. Especially, #6036 and  #5114.

**Comment by jstrachan** [comment]
@ccojocar when using github apps there is not a single user + token for github.com; but one per owner (organisation/user). So we need to be able to generate the git credentials dynamically based on the owner of a github repository.

It's easy to patch tekton to do this; then the generated `git/credentials` file is specific to a github owner. 

However unless we have this patch to be able to load `git/credentials` we've other easy way to align the credentials that tekton creates to the ones used by `jx` steps if using github apps.

Also since all steps in a tekton pipeline have access to the generated git credentials; we could consider removing all the custom logic in `jx` to load `Secrets` if inside a pipeline; and rely on getting secrets from the local `git/credentials` - whether running locally or inside a tekton pipeline; further aligning and simplifying things

**Comment by ccojocar** [comment]
Thanks for clarifications!

>However unless we have this patch to be able to load git/credentials we've no easy way to align the credentials that tekton creates to the ones used by jx steps if using github apps (which loads them from ~/.jx/gitAuth.yaml or Secrets directly).

This is not always the case. The gitAuth.yaml is stored in `vault` when vault is enabled. I am wondering why we are not storing those GitHub app generated credentials in vault.

**Comment by ccojocar** [comment]
I am working on moving all these auth config handlers behind the `AuthConfigService`. It would be nice at least to implement the `ConfigHandler` interface, if you want to load them form that git credentials file. Just have a look how is done in `file_config_handler.go`, and all other `*_handler.go` sources.  Then in the factory, we go over all these handler and load the credentials either based on the configured sources, or lookup them in different places. 

This is to avoid having commands which loads directly the pipeline secrets from kubernetes secrets as of today. The plan is to have this behind this service abstraction.

**Comment by jstrachan** [comment]
@ccojocar yeah, should be able to just merge this logic into the `file_config_handler.go`.

If this approach works (using the `git/credentials` from tekton) we could look at disabling the logic of looking at the secrets at all if folks are using tekton.

**Comment by ccojocar** [comment]
It would be great if you could merge them in the file handler.

**Comment by jstrachan** [comment]
@ccojocar actually on reviewing the change - we'll need this logic for file, configmap and secret based auth implementations; so the position of this change looks correct to me

**Comment by ccojocar** [comment]
We partially use this logic to create the auth config service (in some commands):https://github.com/jenkins-x/jx/blob/4a89356b28368bb9724d55fbf44c1c9882f66d2f/pkg/cmd/clients/factory.go#L368, while other commands loads the auth directly from kubernetes secrets see `LoadPipelineSecrets` - which is not quite right and leads to quite some issues (e.g. when vault is enabled). 

The plan is to move everything under this factory method (abstraction) and based on secrets location config and context, we will create the git auth config service in one place. 

It think you can modify the file handler such that:
- first try to load the gitAuth.yaml file form ~/.jx
- second tries to parse the credentials form `.git/credentials` file

**Comment by jstrachan** [comment]
@ccojocar thanks Cosmin, I'll do that.

BTW on further testing of the use of GitHub App tokens and tekton; it turns out that loading git/credentials is great for working with git & promotes consistency with tekton and avoids us needing `jx step git credentials`) - we do still need to load tokens from Secrets. e.g. `jx promote` can often end up using a different git owner and token than the application we are releasing/promoting. So I don't think we can zap the logic to load secrets from Secrets (or vault) any time soon; but we can hopefully make `jx step git credentials` become a no-op when using tekton.

**Comment by jstrachan** [comment]
@ccojocar I moved the git credentials loading inside the file config handler: https://github.com/jenkins-x/jx/blob/a550d97bd1695103022ccd0c360a9af56216159e/pkg/auth/file_config_handler.go#L59

**Pull request merged**